invokeai/app/services/config/config_default.py

```diff
@@ -225,397 +47,383 @@ def validate_url_regex(cls, v: str) -> str:
+                parsed_config_dict["vram"] = v
+            if k == "conf_path":
+                parsed_config_dict["legacy_models_yaml_path"] = v
+            elif k in InvokeAIAppConfig.model_fields:
+                # skip unknown fields
+                parsed_config_dict[k] = v
+    config = InvokeAIAppConfig.model_validate(parsed_config_dict)
+
+    return config
+
+
+def load_and_migrate_config(config_path: Path) -> InvokeAIAppConfig:
+    """Load and migrate a config file to the latest version.
+
+    Args:
+        config_path: Path to the config file.
+
+    Returns:
+        An instance of `InvokeAIAppConfig` with the loaded and migrated settings.
+    """
+    assert config_path.suffix == ".yaml"
+    with open(config_path) as file:
+        loaded_config_dict = yaml.safe_load(file)
 
-    @property
-    def use_cpu(self) -> bool:
-        """Return true if the device is set to CPU or the always_use_cpu flag is set."""
-        return self.always_use_cpu or self.device == "cpu"
+    assert isinstance(loaded_config_dict, dict)
 
-    @property
-    def disable_xformers(self) -> bool:
-        """Return true if enable_xformers is false (reversed logic) and attention type is not set to xformers."""
-        disabled_in_config = not self.xformers_enabled
-        return disabled_in_config and self.attention_type != "xformers"
+    if "InvokeAI" in loaded_config_dict:
+        # This is a v3 config file, attempt to migrate it
+        shutil.copy(config_path, config_path.with_suffix(".yaml.bak"))
+        try:
+            config = migrate_v3_config_dict(loaded_config_dict)
+        except Exception as e:
+            shutil.copy(config_path.with_suffix(".yaml.bak"), config_path)
+            raise RuntimeError(f"Failed to load and migrate v3 config file {config_path}: {e}") from e
+        # By excluding defaults, we ensure that the new config file only contains the settings that were explicitly set
+        config.write_file(config_path)
+        return config
+    else:
+        # Attempt to load as a v4 config file
+        try:
+            # Meta is not included in the model fields, so we need to validate it separately
+            config = InvokeAIAppConfig.model_validate(loaded_config_dict)
+            assert (
+                config.schema_version == CONFIG_SCHEMA_VERSION
+            ), f"Invalid schema version, expected {CONFIG_SCHEMA_VERSION}: {config.schema_version}"
+            return config
+        except Exception as e:
+            raise RuntimeError(f"Failed to load config file {config_path}: {e}") from e
 
-    @property
-    def profiles_path(self) -> Path:
-        """Path to the graph profiles directory."""
-        return self._resolve(self.profiles_dir)
 
-    @staticmethod
-    def find_root() -> Path:
-        """Choose the runtime root directory when not specified on command line or init file."""
-        return _find_root()
+@lru_cache(maxsize=1)
```

Comment:
This is such a neat way to managing singletons. Cool that it passes around the object instead of copies of it too. Wasn't sure if this would keep the changes on the second call but it did.
```
app_config = get_config()
app_config.log_level = "warning"
app_config = get_config()
```

Reply:
Yeah, nifty trick :)